src/OpenTracing/TracerSpanBuilderExtensions.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace OpenTracing
+{
+    using System.Linq;
+
+    public static class TracerSpanBuilderExtensions
+    {
+        public static ISpanBuilder BuildSpan(ITracer tracer, string operationName)
+        {
+            return new SpanBuilder(tracer, operationName);
+        }
+
+        private sealed class SpanBuilder : ISpanBuilder
+        {
+            private readonly ITracer tracer;
+
+            private string operationName;
+            private ICollection<Tuple<string, ISpanContext>> references;
+            private DateTimeOffset? explicitStartTime;
+            private IDictionary<string, object> tags;
```

Comment:
The second commit in this PR introduced this unified dictionary in place of individual collections. This is likely less CPU performant due to boxing, but makes for more clear code.

An alternative but relatively equal change would be to keep the separate collections and a set of known keys (to ensure duplicate keys with different values are handled properly).